src/modules/python/pbs_hooks/PBS_xeon_phi_provision.PY

```diff
@@ -0,0 +1,72 @@
+import pbs
+import sys
+import os
+import time
+
+sys.path.append('/opt/pbs/python/lib/python2.7')
+import subprocess
+import json as json
+
+# Section to be moved to config file once available for provision hook
+capmc_dir = '/opt/cray/capmc/default/bin'
+power_down_checks = 10
+power_down_sleep = 15
+power_up_checks = 20
+power_up_sleep = 60
+
+# Get the information about the node and aoe to provision
+e = pbs.event()
+vnode = e.vnode
+aoe = e.aoe
+
+numa_cfg, cache_percent = aoe.split('_')
+nid = vnode.split('_')[1]
+
+# Add capmc_dir to path
+os.environ['PATH'] = capmc_dir + ":" + os.environ['PATH']
+
+# capmc commands to configure the knl nodes
+cmd_set_numa_cfg = 'capmc set_numa_cfg --nids %s --mode %s' % (nid, numa_cfg)
+cmd_set_mcdram_cfg = 'capmc set_mcdram_cfg --nids %s --mode %s' % (
+    nid, cache_percent)
+cmd_node_reinit = 'capmc node_reinit --nids %s' % (nid)
+cmd_node_status = ['capmc', 'node_status', '--nids', '%s' % (nid)]
+
+pbs.logmsg(pbs.EVENT_DEBUG3, "vnode: %s" % vnode)
+pbs.logmsg(pbs.EVENT_DEBUG3, "aoe: %s" % aoe)
+pbs.logmsg(pbs.EVENT_DEBUG3, "numa cmd: %s" % cmd_set_numa_cfg)
+pbs.logmsg(pbs.EVENT_DEBUG3, "cache cmd: %s" % cmd_set_mcdram_cfg)
+pbs.logmsg(pbs.EVENT_DEBUG3, "power reinit cmd: %s" % cmd_node_reinit)
+
+appret1 = os.system(cmd_set_numa_cfg)
+pbs.logmsg(pbs.EVENT_DEBUG3, "appret1: %s" % str(appret1))
+appret2 = os.system(cmd_set_mcdram_cfg)
+pbs.logmsg(pbs.EVENT_DEBUG3, "appret2: %s" % str(appret2))
+appret3 = os.system(cmd_node_reinit)
+pbs.logmsg(pbs.EVENT_DEBUG3, "appret3: %s" % str(appret3))
+
+# Nodes can take upwards of 15 minutes to come back online
+poll_val = 0
+poll_cnt = 0
+while poll_val < 1 and poll_cnt < power_up_checks:
+    time.sleep(power_up_sleep)
+    process = subprocess.Popen(
+        cmd_node_status,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE)
+    output, err = process.communicate()
+    node_status = json.loads(output)
+    pbs.logmsg(
+        pbs.EVENT_DEBUG3, "poll_cnt: %d node_status: %s" %
+        (poll_cnt, node_status))
+    if "ready" in node_status:
+        if int(nid) in node_status['ready']:
+            pbs.logmsg(pbs.EVENT_DEBUG3, "Node was successfully powered on")
+            poll_val = 1
+    poll_cnt += 1
+
+if poll_val != 1:
+    e.reject("Provisioning with reboot failed", 211)
```

Comment:
Is there a special reason to give 211?

Reply:
This number to identify from where reject is called, otherwise by default reject returns 255.